### ISSUE TITLE
protoprint generates duplicated proto options when a proto option type is repeated

### ISSUE THREAD

**Issue by yb-zhu** [jhump/protoreflect#363]
I use protoprint to get proto source code from proto descriptor set. After recently protoreflect library changes, I start to see duplicated proto options in the generated source code when a proto option type is repeated.

In the  

```
message Foo {
  option (foo.options.qux) = "Message level option";
  **option (foo.options.repeated_string) = "Group_A";  // The line and the line below are duplicated
  option (foo.options.repeated_string) = "Group_A";** 
  ...
}
```

Where foo.options.repeated_string is defined as 

```
extend proto2.MessageOptions {
  repeated string repeated_string = 123456789;
}
```

**Comment by jhump** [comment]
@yb-zhu, can you provide more info on the actual descriptor you were printing? Are you sure the descriptor did not have two entries in the repeated field, both with the same value? There are tests for this package that do exercise various cases for repeated options, including cases where there is just one vs. multiple values. And all tests pass.

**Comment by yb-zhu** [comment]
@jhump Thanks for the reply.  I built a descriptor set from a proto message just like what you did for protoreflect/internal/testprotos/desc_test_comments.proto. Then I extract the proto descriptor and call protoprint to reproduce the source code.  In my input proto file, I only have a single entry in the repeated field. 

Can you try to add a proto option of a repeated type to the message in protoreflect/internal/testprotos/desc_test_comments.proto? And then run protoreflect/desc/protoprint/print_test.go. I think that will reproduce the issue.

**Comment by jhump** [comment]
> Can you try to add a proto option of a repeated type to the message in protoreflect/internal/testprotos/desc_test_comments.proto?

@yb-zhu, as I already stated, I _already have_ a test case that exercises this.

[This option](https://github.com/jhump/protoreflect/blob/master/internal/testprotos/desc_test_complex.proto#L83) is repeated and is used with a single value [here](https://github.com/jhump/protoreflect/blob/master/internal/testprotos/desc_test_complex.proto#L51) and [here](https://github.com/jhump/protoreflect/blob/master/internal/testprotos/desc_test_complex.proto#L53) (and [here](https://github.com/jhump/protoreflect/blob/master/internal/testprotos/desc_test_complex.proto#L90) is a case with multiple values, just for completeness). And the `protoprint` package correctly handles these just fine as you can see from test data  [here](https://github.com/jhump/protoreflect/blob/master/desc/protoprint/testfiles/desc_test_complex-default.proto#L64) and [here](https://github.com/jhump/protoreflect/blob/master/desc/protoprint/testfiles/desc_test_complex-default.proto#L67) (and [here](https://github.com/jhump/protoreflect/blob/master/desc/protoprint/testfiles/desc_test_complex-default.proto#L108) is the output for multiple values).

This scenario is exactly as you describe but does **not** reproduce the issue.

That's why I requested _working_ example code from you that reproduces what you are seeing.

**Comment by aberasarte** [comment]
I think that I got the exact same issue after upgrading to 1.7.0 (I was in 1.5.0 before).

I've been able to reproduce it with the [desc_test_complex.proto](https://github.com/jhump/protoreflect/blob/93246e23ccdf09f2739a50e980e5e9188fc67a90/internal/testprotos/desc_test_complex.proto) as well. If I parse that proto with the `protoparser` and then print it again with the `protoprinter`, this is the output I get for the `TestService`:

```
service TestTestService {
  rpc Get ( Test ) returns ( Test ) {
    option (validator) = { authenticated:true permission:<action:READ entity:"user"> permission:<action:READ entity:"user"> };
  }

  rpc UserAuth ( Test ) returns ( Test ) {
    option (validator) = { authenticated:true permission:<action:LOGIN entity:"client"> permission:<action:LOGIN entity:"client"> };
  }
}
```

The permission option is printed twice for each method.

**Comment by jhump** [comment]
@aberasarte, thank you for the info. Can you provide a code example that reproduces the issue?

Like I pointed out before, this situation is already covered in unit tests and it produces [this output](https://github.com/jhump/protoreflect/blob/master/desc/protoprint/testfiles/desc_test_complex-default.proto#L149), which shows only one instance of the option. So I need an actual code sample that demonstrates the issue (since that code clearly does something different than the test I've already written for this case).

**Comment by aberasarte** [comment]
> Can you provide a code example that reproduces the issue?

Here you are https://github.com/aberasarte/protoreflect-test. Running `go run main.go` it prints:

```
syntax = "proto2";

package foo.bar;

import "google/protobuf/descriptor.proto";

option go_package = "github.com/aberasarte/protoreflect-test/";

message Test {
}

message Validator {
  optional bool authenticated = 1;

  repeated Permission permission = 2;

  message Permission {
    optional Action action = 1;

    optional string entity = 2;
  }

  enum Action {
    LOGIN = 0;

    READ = 1;

    WRITE = 2;
  }
}

service TestTestService {
  rpc Get ( Test ) returns ( Test ) {
    option (validator) = { authenticated:true permission:<action:READ entity:"user"> permission:<action:READ entity:"user"> };
  }

  rpc UserAuth ( Test ) returns ( Test ) {
    option (validator) = { authenticated:true permission:<action:LOGIN entity:"client"> permission:<action:LOGIN entity:"client"> };
  }
}

extend google.protobuf.MethodOptions {
  optional Validator validator = 12345;
}
```

I hope it helps.

**Comment by aberasarte** [comment]
BTW, if I run the code from VSCode in debug mode it panics. I guess that this is what you were talking about in https://github.com/jhump/protoreflect/issues/341#issuecomment-662799603. I have both the IDE and the Go tools updated, I don't know if both issues could be somehow related.

**Comment by jhump** [comment]
@aberasarte, thank you for the repro case! It is alarming how simple it is and that the tests I already have don't encounter this.

I dug into it a bit. In fact, when I first pasted your code _into_ my tests, they still worked fine! After a little more exploring, I was able to isolate the difference between my tests working and your example not working: this one line in your `go.mod` file:
```
	google.golang.org/protobuf v1.25.0 // indirect
```
If you remove that, the problem goes away!

However, what is _really_ weird is the problem returns if you do `go get github.com/jhump/protoreflect@master`. In that newer revision, this repo _does_ use that library, and uses v1.25 (actually an unreleased commit after v1.25). And the failure is reproduced. _However_, on master in my repo, my tests all still pass. Even when I paste in the same code as your repro case into a test, it works fine and does not double-print. Super weird!

I need to dig more to figure out what exactly is going on.

**Comment by jhump** [comment]
Alright, I've figured it out.

1. This was introduced by using v1.4 or higher of "github.com/golang/protobuf".
    * The new representation of messages stores _all_ unrecognized fields together. However in v1.3, unrecognized _extensions_ were separate from other unrecognized fields.
    * So the code in the `dynamic` package in this repo, that is trying to interpret unrecognized fields, finds them twice: it sees them when it is examining unrecognized extensions and _again_ when it is examining other unrecognized fields.
    * This does not impact non-repeated fields since they simple keep their last observed value. But, for repeated fields, all values are concatenated together.
2. My tests are immune to this because... the options I am testing are _not_ unknown.
    * Since my test also links in the "internal/testprotos" package, it links in the resulting generated Go code for these protos. So because they _recognized_, they don't fall into the trap described above, and only print once.

So I can add a repro case by having a test that parses a proto that is _not_ also an imported Go package.

I also see the fix: in v1.3, unknown extensions will _never_ be part of other unrecognized fields. But in v1.4, unknown extensions are _always_ part of the other unrecognized fields. So the code needs to track the tag numbers it sees in the unrecognized fields and _discard_ any unknown extensions with a matching tag.

**Comment by jhump** [comment]
This issue is fixed in v1.7.1 and will soon have a fix in master, too (and thus also in the upcoming v1.8.0 release).

**Comment by aberasarte** [comment]
That was fast! I tested it and I can confirm that it is working for my use case as well. Thank you very much!

> The new representation of messages stores all unrecognized fields together. However in v1.3, unrecognized extensions were separate from other unrecognized fields.

Oh yes, I forgot to mention that I had to set `github.com/golang/protobuf` (which is deprecated) to v1.3 as a workaround. With the fix, I can use `google.golang.org/protobuf` and make our linter happy 👏 .
